### PR TITLE
MFA with TOTP

### PR DESCRIPTION
**Pull request by louischan-oursky**
ref #56

### CODE REVIEW

File: pkg/auth/dependency/newinteraction/nodes/create_authenticator_totp_setup.go [hunk 1]
```diff
@@ -10,26 +10,23 @@ func init() {
...
 	Stage newinteraction.AuthenticationStage
 }
 
 func (e *EdgeCreateAuthenticatorTOTPSetup) Instantiate(ctx *newinteraction.Context, graph *newinteraction.Graph, rawInput interface{}) (newinteraction.Node, error) {
-	_, ok := rawInput.(InputCreateAuthenticatorTOTPSetup)
-	if !ok {
-		return nil, newinteraction.ErrIncompatibleInput
-	}
-
+	// This edge does not take any input so it always instantiate.
```
[the remark this answers is not in the document]
Comment (louischan-oursky) [reply pkg/auth/dependency/newinteraction/nodes/create_authenticator_totp_setup.go:18]: Yes. Let me add a TODO here.

File: pkg/auth/handler/webapp/setup_totp.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+			result, err := h.WebApp.PostInput(StateID(r), func() (input interface{}, err error) {
+				err = SetupTOTPSchema.PartValidator(SetupTOTPRequestSchema).ValidateValue(FormToJSON(r.Form))
+				if err != nil {
+					return
+				}
+
+				now := h.Clock.NowUTC()
+
+				// FIXME(mfa): decide a proper display name.
+				displayName := fmt.Sprintf("TOTP @ %s", now.Format(time.RFC3339))
```
[the remark this answers is not in the document]
Comment (louischan-oursky) [reply pkg/auth/handler/webapp/setup_totp.go:193]: The design does not have such input field. And I never see it when I am using other online service, such as AWS and Google. So it should be generated by us somehow.

[the remark this answers is not in the document]
Comment (louischan-oursky) [reply pkg/auth/handler/webapp/setup_totp.go:193]: Instead the user may be able to change the display later on in the settings page. Requiring the user to think of a display name may cause friction.

[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/auth/handler/webapp/setup_totp.go:193]: Let's just keep the placeholder for now.

**Pull request merged**